### PR TITLE
[luxon] Upgrade to 2.0

### PR DESCRIPTION
**Pull request by hugofpsilva**
Please fill in this template.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://moment.github.io/luxon/#/upgrading
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

Had to add a // tslint:disable-next-line:no-unnecessary-class because the Settings class implemented by package author only has static methods.

### CODE REVIEW

**Comment by peterblazejewicz** [comment]
This is popular package and this change will limit usage of it to TS 4.3 or newer (so effectively current edge version). What's the reason behind? One can use `readonly prop: type;` for getters instead of relying on version support or can split definition into version branches (whihc would be an overkill IMO)

**Comment by hugofpsilva** [comment]
@peterblazejewicz One property of the Settings class has a different type getter and setter, which caused an error in the tests which I "fixed" using the recommended action, adding a minimum TS version.

I now deleted the minimum TS version, and broadened the property's type to work both as setter and getter. It's not perfect, but I believe it's an OK solution until this can be bumped to TS 4.3.

File: types/luxon/src/datetime.d.ts [hunk 1]
```diff
@@ -171,947 +162,1086 @@ export interface ExplainedFormat {
...
+     *
+     * @example DateTime.now().toObject() //=> { year: 2017, month: 4, day: 22, hour: 20, minute: 49, second: 42, millisecond: 268 }
      */
-    hasSame(other: DateTime, unit: DurationUnit): boolean;
+    toObject(options?: {
+        /**
+         * Include configuration attributes in the output
+         * @defaultValue false
+         */
+        includeConfig?: boolean;
```
Comment (peterblazejewicz) [inline types/luxon/src/datetime.d.ts:1000]: is this supposed to be there?:
https://moment.github.io/luxon/#/upgrading?id=other-breaking-changes
> DateTime#toObject no longer accepts an includeConfig option

**Review by CarsonF — COMMENTED** [review]
Can you add the line break back to all the example tags? The text following on the same line is for the example's caption and is not formatted as TS (at least that's how WebStorm does it)

**Comment by peterblazejewicz** [comment]
the changes like:
```ts
locale?: string | undefined;
```
were added by DT maintainer for compatiblity with strict modes coming with 4.4. That's transparent for Luxon itself. Do we have a reason to remove those changes?

**Comment by CarsonF** [comment]
Reordering all of the members makes it really hard to see what's changed.

File: types/luxon/src/datetime.d.ts (hunk 1 above)
Comment (hugofpsilva) [reply types/luxon/src/datetime.d.ts:1000]: I believe it is a mistake on the upgrading guide.
includeConfig is still accepted in 2.0.1 and setting it true changes the object returned.
It is also in the API docs for DateTime. See https://moment.github.io/luxon/api-docs/index.html#datetimetoobject

**Comment by hugofpsilva** [comment]
> Can you add the line break back to all the example tags? The text following on the same line is for the example's caption and is not formatted as TS (at least that's how WebStorm does it)

> Reordering all of the members makes it really hard to see what's changed.

@CarsonF due to the large amount of changes from the previous version I restarted the typings from the author's code. The typings are now ordered in line with the author's code for future simplicity, but that meant copying the comments as well and those example tags came wrongly formatted from source. I will correct that in the next commit.

**Comment by hugofpsilva** [comment]
> the changes like:
> 
> ```ts
> locale?: string | undefined;
> ```
> 
> were added by DT maintainer for compatiblity with strict modes coming with 4.4. That's transparent for Luxon itself. Do we have a reason to remove those changes?

I don't, I didn't see it was for a future TS version, I will add those back. Thanks!

**Comment by peterblazejewicz** [comment]
@hugofpsilva can we first review this #54791, as being shorter? 
Your changes are more complex, including changes to how code is used (proper getters, exposed on class instance, etc).

**Comment by hugofpsilva** [comment]
@peterblazejewicz  Absolutely. Should I close this one and then reopen after, if the changes are still worth doing so?

**Comment by peterblazejewicz** [comment]
You can rebase, you don't need to close, when the other is merged. Whatever is easier for you, pick one, thx!

**Comment by peterblazejewicz** [comment]
I think we could alias `DurationInput` (and deprecate) and just follow source code naming.
`Duration` | `DurationObject` | `number` are `durationish`. Convention for this is `DurationLike`.
@hugofpsilva thoughts?

**Comment by hugofpsilva** [comment]
@peterblazejewicz Updated the DurationInput type to DurationLike as you suggested.

**Comment by peterblazejewicz** [comment]
Let me kindly ask @CarsonF for yet one review, thx!

**Review by CarsonF — APPROVED** [review]
LGTM

**Review by peterblazejewicz — APPROVED** [review]

**Pull request merged**